[QA] Move integration tests to vizro

l0uden opened the pull request:
## Description

Here's the first iteration of tests moving from vizro-qa to vizro repo.
Here I moved (and refactored) part of the default dashboard, test helpers and tests for filters, pages and themes. Also created CI for running the tests
The aim of this PR is to find the way of organising this tests in the proper way, so I can move the rest of them after in the same manner

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

    - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
    - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
    - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
    - I have not referenced individuals, products or companies in any commits, directly or indirectly.
    - I have not added data or restricted code in any commits, directly or indirectly.

petar-qb reviewed (commented):
This looks great to me so far! 🚀 Amazing work, @l0uden! 🥇

That said, I think we should discuss a few key topics on a call rather than communicating everything through PR comments. Moving all the tests from the `vizro-qa` repo to the `vizro` repo is a significant change and potentially introduces some risks (like unoptimised folder structure...). It’s also a great opportunity to optimise while changes are still manageable and cheap because when the full migration is complete, making similar updates could become much more complex. 

What do you think about this @antonymilne @Joseph-Perkins?
To reduce risks and streamline the process, I propose setting up the call about the Vizro tests architecture. Addressing these points now will save time and effort later, and it will help us to merge this PR faster (than communicating these things over the PR comments), allowing @l0uden to proceed with the test migration smoothly.

Some topics/questions I'd like to discuss:
1. Should we add a dedicated `test` hatch environment?
2. Can we improve the folder structure for better maintainability?
3. Should we combine `vizro-core/tests/integrations` with `vizro-core/tests/e2e` (and apply the same approach to `vizro-ai`)?

l0uden commented:
> This looks great to me so far! 🚀 Amazing work, @l0uden! 🥇
> 
> That said, I think we should discuss a few key topics on a call rather than communicating everything through PR comments. Moving all the tests from the `vizro-qa` repo to the `vizro` repo is a significant change and potentially introduces some risks (like unoptimised folder structure...). It’s also a great opportunity to optimise while changes are still manageable and cheap because when the full migration is complete, making similar updates could become much more complex.
> 
> What do you think about this @antonymilne @Joseph-Perkins? To reduce risks and streamline the process, I propose setting up the call about the Vizro tests architecture. Addressing these points now will save time and effort later, and it will help us to merge this PR faster (than communicating these things over the PR comments), allowing @l0uden to proceed with the test migration smoothly.
> 
> Some topics/questions I'd like to discuss:
> 
> 1. Should we add a dedicated `test` hatch environment?
> 2. Can we improve the folder structure for better maintainability?
> 3. Should we combine `vizro-core/tests/integrations` with `vizro-core/tests/e2e` (and apply the same approach to `vizro-ai`)?

Thanks @petar-qb ! 

Completely agree with you about the call about this, let's discuss time on Monday, when @antonymilne will be available. And also it would be better if he could have time to take a look on what's done.

maxschulz-COL commented on vizro-core/tests/tests_utils/e2e/checkers.py:
I would argue a one liner function description could help some of these tests?

maxschulz-COL commented on vizro-core/tests/e2e/vizro/dashboards/default/dashboard_pages.py:
did you ever think of splitting this up into topical categories? Like test_filters, test_parameters, etc.

Not saying we need to do this 100%, just an idea. The files could have a description of broadly what functionality the tests in the file cover

maxschulz-COL replied:
Or after reading further in your PR, could this make sense to match the files in `test_dom_elements`. I guess there are the actual tests, whereas the functions/files here are the basis for it. Just an idea!

maxschulz-COL commented on vizro-core/tests/e2e/vizro/dashboards/wait-for-it.sh:
Is this the right location for this file? Is that not a utils?

maxschulz-COL commented on vizro-core/tests/tests_utils/e2e/custom_feature_helpers/custom_actions/custom_actions.py:
I am not sure the custom feature helpers are something that should be in utils, is that not rather something that is (like dashboards) supposed to be where it is used? I assume here that these are only used in e2e tests

maxschulz-COL replied:
This applies to all files in the `custom_feature_helpers` folder

maxschulz-COL commented on vizro-core/tests/tests_utils/e2e/helpers.py:
Same comment for some functions in these files (not all!) - a one liner could help

maxschulz-COL reviewed (approved):
⭐ Looks good to me 🚀 

I have some suggestions, please only implement them if they seem useful, and they make sense!

petar-qb commented:
Considering all that @maxschulz-COL suggested and everything we aligned on during last week’s call, I’ve prepared the following folder structure proposal. This structure closely resembles what we currently have but introduces the following changes:

1. Renames `e2e/vizro/dashboards/` -> `e2e/vizro/component_utils/`
2. Moves `tests_utils/e2e/custom_feature_helpers/` under `e2e/vizro/custom_components/`
3. Extracts `e2e/vizro/dashboards/default/dashboard.py` to a higher-level directory: `e2e/vizro/dashboards/dashboard.py`
4. Breaks `e2e/vizro/dashboards/default/dashboard_pages.py` into multiple files and renames the parent folder from `default/` to `pages/`
5. Defines a new structure of `e2e/vizro/test_screenshots/`
6. Move all 6 files from `tests_utils/e2e/` into the new `e2e/vizro/utils/`
7. Renames `helpers.py` -> `waiters.py`

```
e2e/
├── component_library/
│   └── test_component_library.py
├── vizro/
│   ├── components_utils/  (1. Renames `e2e/vizro/dashboards/` -> `e2e/vizro/component_utils/`)
│   │   ├── assets/
│   │   │   └── ...
│   ├── custom_components/  (2. Moves `tests_utils/e2e/custom_feature_helpers/` under `e2e/vizro/custom_components/`)
│   │   ├── custom_actions/
│   │   │   └── ...
│   │   ├── custom_charts/
│   │   │   └── ...
│   │   ├── custom_components/
│   │   │   └── ...
│   ├── dashboard.py  (3. Extracts `e2e/vizro/dashboards/default/dashboard.py` to a higher-level directory: `e2e/vizro/dashboards/dashboard.py`)
│   ├── pages/  (4. Breaks `e2e/vizro/dashboards/default/dashboard_pages.py` into multiple files and renames the parent folder from `default/` to `pages/`)
│   │   ├── filter_page/
│   │   ├── parameter_page.py
│   │   └── ...
│   ├── test_dom_elements/
│   │   ├── conftest.py
│   │   ├── test_charts.py
│   │   ├── test_filters.py
│   │   ├── test_pages.py
│   │   └── ...
│   ├── test_screenshots/  (5. Defines a new structure of `e2e/vizro/test_screenshots/`)
│   │   ├── screenshot_images/
│   │   │   └── ... 
│   │   └── test_screenshots.py
│   ├── utils/  (6. Move all 6 files from `tests_utils/e2e/` into the new `e2e/vizro/utils/`)
│   │   ├── asserts.py
│   │   ├── checkers.py
│   │   ├── constants.py
│   │   ├── navigation.py
│   │   ├── paths.py
│   │   └── waiters.py  (7. Renames `helpers.py` -> `waiters.py`)
tests_utils/
...
```

@maxschulz-COL @l0uden @antonymilne what do you think about the seven changes I suggested?

maxschulz-COL commented:
> Considering all that @maxschulz-COL suggested and everything we aligned on during last week’s call, I’ve prepared the following folder structure proposal. This structure closely resembles what we currently have but introduces the following changes:
> 
> 1. Renames `e2e/vizro/dashboards/` -> `e2e/vizro/component_utils/`
> 2. Moves `tests_utils/e2e/custom_feature_helpers/` under `e2e/vizro/custom_components/`
> 3. Extracts `e2e/vizro/dashboards/default/dashboard.py` to a higher-level directory: `e2e/vizro/dashboards/dashboard.py`
> 4. Breaks `e2e/vizro/dashboards/default/dashboard_pages.py` into multiple files and renames the parent folder from `default/` to `pages/`
> 5. Defines a new structure of `e2e/vizro/test_screenshots/`
> 6. Move all 6 files from `tests_utils/e2e/` into the new `e2e/vizro/utils/`
> 7. Renames `helpers.py` -> `waiters.py`
> 
> ```
> e2e/
> ├── component_library/
> │   └── test_component_library.py
> ├── vizro/
> │   ├── components_utils/  (1. Renames `e2e/vizro/dashboards/` -> `e2e/vizro/component_utils/`)
> │   │   ├── assets/
> │   │   │   └── ...
> │   ├── custom_components/  (2. Moves `tests_utils/e2e/custom_feature_helpers/` under `e2e/vizro/custom_components/`)
> │   │   ├── custom_actions/
> │   │   │   └── ...
> │   │   ├── custom_charts/
> │   │   │   └── ...
> │   │   ├── custom_components/
> │   │   │   └── ...
> │   ├── dashboard.py  (3. Extracts `e2e/vizro/dashboards/default/dashboard.py` to a higher-level directory: `e2e/vizro/dashboards/dashboard.py`)
> │   ├── pages/  (4. Breaks `e2e/vizro/dashboards/default/dashboard_pages.py` into multiple files and renames the parent folder from `default/` to `pages/`)
> │   │   ├── filter_page/
> │   │   ├── parameter_page.py
> │   │   └── ...
> │   ├── test_dom_elements/
> │   │   ├── conftest.py
> │   │   ├── test_charts.py
> │   │   ├── test_filters.py
> │   │   ├── test_pages.py
> │   │   └── ...
> │   ├── test_screenshots/  (5. Defines a new structure of `e2e/vizro/test_screenshots/`)
> │   │   ├── screenshot_images/
> │   │   │   └── ... 
> │   │   └── test_screenshots.py
> │   ├── utils/  (6. Move all 6 files from `tests_utils/e2e/` into the new `e2e/vizro/utils/`)
> │   │   ├── asserts.py
> │   │   ├── checkers.py
> │   │   ├── constants.py
> │   │   ├── navigation.py
> │   │   ├── paths.py
> │   │   └── waiters.py  (7. Renames `helpers.py` -> `waiters.py`)
> tests_utils/
> ...
> ```
> 
> @maxschulz-COL @l0uden @antonymilne what do you think about the seven changes I suggested?

That sounds good to me! Thanks @petar-qb! All perfect except maybe `dashboard.py`. Could it be `dashboard_for_testing.py`? (super minor though, feel free to also ignore)

petar-qb reviewed (approved):
This looks really good now.  🚀
